In [2]:
import seaborn as sns
import matplotlib.pyplot as plt
import time
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

import keras
from keras.models import Sequential
from keras.layers import *
from keras.optimizers import Adam
from keras.layers import Activation
from keras.callbacks import EarlyStopping, ModelCheckpoint
from sklearn.decomposition import PCA

import matplotlib.pyplot as plt



In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
train = pd.read_csv('/content/drive/Shareddrives/TNT/스터디/2021년 겨울방학/과제/5주차 과제/data/train.csv')
test = pd.read_csv('/content/drive/Shareddrives/TNT/스터디/2021년 겨울방학/과제/5주차 과제/data/test.csv')

print(train.columns)
print(test.columns)

Index(['layer_1', 'layer_2', 'layer_3', 'layer_4', '0', '1', '2', '3', '4',
       '5',
       ...
       '216', '217', '218', '219', '220', '221', '222', '223', '224', '225'],
      dtype='object', length=230)
Index(['id', '0', '1', '2', '3', '4', '5', '6', '7', '8',
       ...
       '216', '217', '218', '219', '220', '221', '222', '223', '224', '225'],
      dtype='object', length=227)


In [5]:
train.head()

layer_1  layer_2  layer_3  layer_4  ...       222       223       224       225
0       10       10       10       10  ...  0.419225  0.443250  0.433414  0.465502
1       10       10       10       20  ...  0.637706  0.625981  0.653231  0.637853
2       10       10       10       30  ...  0.747743  0.746037  0.737356  0.750391
3       10       10       10       40  ...  0.774712  0.801526  0.805305  0.784057
4       10       10       10       50  ...  0.787763  0.794948  0.819105  0.801781

[5 rows x 230 columns]

In [6]:
test.head()

id         0         1         2  ...       222       223       224       225
0   0  0.535410  0.520775  0.494087  ...  0.738704  0.739460  0.702139  0.702238
1   1  0.351099  0.398179  0.413809  ...  0.099780  0.145420  0.260501  0.343857
2   2  0.490537  0.435958  0.413428  ...  0.845859  0.846032  0.836724  0.846779
3   3  0.051634  0.075802  0.133983  ...  0.382823  0.489381  0.562383  0.599247
4   4  0.154031  0.201728  0.270414  ...  0.530573  0.582752  0.637296  0.637238

[5 rows x 227 columns]

In [7]:
print(train.shape)
print(test.shape)

(810000, 230)
(10000, 227)


train 데이터는 810000개의 레코드와 230개의 컬럼으로 이루어져있음   
test 데이터는 10000개의 레코드와 227개의 컬럼으로 이루어져 있음.  

이때, test 데이터에는 id 값을 제외하고 사용해야함. 

---

In [8]:
print(train['layer_1'].value_counts())
print('-----------------------------')
print(train['layer_2'].value_counts())
print('-----------------------------')
print(train['layer_3'].value_counts())
print('-----------------------------')
print(train['layer_4'].value_counts())
print('-----------------------------')

300    27000
290    27000
20     27000
30     27000
40     27000
50     27000
60     27000
70     27000
80     27000
90     27000
100    27000
110    27000
120    27000
130    27000
140    27000
150    27000
160    27000
170    27000
180    27000
190    27000
200    27000
210    27000
220    27000
230    27000
240    27000
250    27000
260    27000
270    27000
280    27000
10     27000
Name: layer_1, dtype: int64
-----------------------------
300    27000
290    27000
20     27000
30     27000
40     27000
50     27000
60     27000
70     27000
80     27000
90     27000
100    27000
110    27000
120    27000
130    27000
140    27000
150    27000
160    27000
170    27000
180    27000
190    27000
200    27000
210    27000
220    27000
230    27000
240    27000
250    27000
260    27000
270    27000
280    27000
10     27000
Name: layer_2, dtype: int64
-----------------------------
300    27000
290    27000
20     27000
30     27000
40     27000
50     27000
60     27000
70     27000


In [9]:
pd.set_option('display.max_columns', 230)

train.describe()

layer_1        layer_2        layer_3        layer_4  \
count  810000.000000  810000.000000  810000.000000  810000.000000   
mean      155.000000     155.000000     155.000000     155.000000   
std        86.554468      86.554468      86.554468      86.554468   
min        10.000000      10.000000      10.000000      10.000000   
25%        80.000000      80.000000      80.000000      80.000000   
50%       155.000000     155.000000     155.000000     155.000000   
75%       230.000000     230.000000     230.000000     230.000000   
max       300.000000     300.000000     300.000000     300.000000   

                   0              1              2              3  \
count  810000.000000  810000.000000  810000.000000  810000.000000   
mean        0.292653       0.292893       0.293125       0.293363   
std         0.181642       0.181857       0.182055       0.182197   
min        -0.014902      -0.014798      -0.014897      -0.014709   
25%         0.135139       0.135180       0.135258       0.135478   
50%         0.286510       0.286874       0.287194       0.287553   
75%         0.435696       0.435956       0.436112       0.436326   
max         0.748205       0.753103       0.749494       0.747389   

                   4              5              6              7  \
count  810000.000000  810000.000000  810000.000000  810000.000000   
mean        0.293666       0.293994       0.294344       0.294792   
std         0.182361       0.182529       0.182693       0.182858   
min        -0.014903      -0.014662      -0.014924      -0.014789   
25%         0.135585       0.135705       0.135858       0.136423   
50%         0.287830       0.288151       0.288705       0.289222   
75%         0.436634       0.437142       0.437395       0.437924   
max         0.748827       0.750392       0.751105       0.749795   

                   8              9             10             11  \
count  810000.000000  810000.000000  810000.000000  810000.000000   
mean        0.295325       0.295757       0.296188       0.296762   
std         0.183057       0.183192       0.183337       0.183532   
min        -0.014775      -0.014945      -0.014818      -0.014792   
25%         0.136704       0.136993       0.137269       0.137756   
50%         0.289905       0.290652       0.291129       0.291843   
75%         0.438657       0.439172       0.439714       0.440647   
max         0.751538       0.750883       0.753622       0.752210   

                  12             13             14             15  \
count  810000.000000  810000.000000  810000.000000  810000.000000   
mean        0.297396       0.298059       0.298745       0.299503   
std         0.183731       0.183932       0.184135       0.184367   
min        -0.014751      -0.014926      -0.014878      -0.014824   
25%         0.138130       0.138556       0.139140       0.139715   
50%         0.292665       0.293451       0.294451       0.295342   
75%         0.441543       0.442668       0.443442       0.444718   
max         0.757742       0.749621       0.756821       0.755842   

                  16             17             18             19  \
count  810000.000000  810000.000000  810000.000000  810000.000000   
mean        0.300220       0.301024       0.301834       0.302585   
std         0.184559       0.184800       0.185021       0.185242   
min        -0.014904      -0.014811      -0.014950      -0.014979   
25%         0.140296       0.140825       0.141515       0.142154   
50%         0.296341       0.297322       0.298261       0.299172   
75%         0.445731       0.447011       0.448062       0.449408   
max         0.759004       0.754317       0.758532       0.759898   

                  20             21             22             23  \
count  810000.000000  810000.000000  810000.000000  810000.000000   
mean        0.303395       0.304331       0.305255       0.306159   
std         0.185466       0.185706       0.185972       0.186212   
min  

---

In [10]:
train.isnull().sum()

layer_1    0
layer_2    0
layer_3    0
layer_4    0
0          0
          ..
221        0
222        0
223        0
224        0
225        0
Length: 230, dtype: int64

In [11]:
test.isnull().sum()

id     0
0      0
1      0
2      0
3      0
      ..
221    0
222    0
223    0
224    0
225    0
Length: 227, dtype: int64

결측치 없음. 

---

---

### 딥러닝 이용

In [12]:
#데이터를 불러옵니다.
import pandas as pd
import warnings

warnings.filterwarnings('ignore') # warning message 안나오도록함.


In [13]:
#독립변수와 종속변수를 분리합니다.
train_X = train.iloc[:,4:]
train_Y = train.iloc[:,0:4]
test_X = test.iloc[:,1:]

In [14]:
adam = keras.optimizers.Adam(lr=0.001, decay=0)
def build_model():
    model = Sequential()  
    model.add(Dense(units=904, activation='relu', input_dim=train_X.shape[1])) 
    model.add(BatchNormalization())
    model.add(Dropout(0.01))
    model.add(Dense(units=452, activation='relu'))   
    model.add(BatchNormalization())
    model.add(Dropout(0.01))
    model.add(Dense(units=339, activation='relu'))   
    model.add(BatchNormalization())
    model.add(Dropout(0.01))
    model.add(Dense(units=226, activation='relu'))   
    model.add(BatchNormalization())
    model.add(Dropout(0.01))
    model.add(Dense(units=113, activation='relu'))   
    model.add(BatchNormalization())
    model.add(Dropout(0.01))
    model.add(Dense(units=4, activation='linear')) 

    model.compile(loss='mae', optimizer=adam, metrics=['mae'])
    return model
    
model = build_model()

In [15]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 904)               205208    
_________________________________________________________________
batch_normalization (BatchNo (None, 904)               3616      
_________________________________________________________________
dropout (Dropout)            (None, 904)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 452)               409060    
_________________________________________________________________
batch_normalization_1 (Batch (None, 452)               1808      
_________________________________________________________________
dropout_1 (Dropout)          (None, 452)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 339)               1

In [ ]:
# 에포크가 끝날 때마다 점(.)을 출력해 훈련 진행 과정을 표시합니다

EPOCHS = 30

history = model.fit(
  train_X, train_Y,
  epochs=EPOCHS, validation_split = 0.2, verbose=0)

In [ ]:
hist = pd.DataFrame(history.history)
hist['epoch'] = history.epoch
hist.tail()


In [ ]:
model = build_model()

early_stop = keras.callbacks.EarlyStopping(monitor='val_loss', patience=10)

history = model.fit(train_X, train_Y, epochs=EPOCHS,
                    validation_split = 0.2, verbose=0, callbacks=early_stop)

print(history)

In [ ]:
loss, mae, mse = model.evaluate(normed_test_data, test_labels, verbose=2)

print("테스트 세트의 평균 절대 오차: {:5.2f} MPG".format(mae))

In [ ]:
#예측값을 생성합니다.
pred_test = model.predict(test_X)

In [ ]:
#submission 파일을 생성합니다.
sample_sub = pd.read_csv('/content/drive/Shareddrives/TNT/스터디/2021년 겨울방학/과제/5주차 과제/data/sample_submission.csv', index_col=0)
submission = sample_sub+pred_test
submission.to_csv('/content/drive/Shareddrives/TNT/스터디/2021년 겨울방학/과제/5주차 과제/data/submission2.csv')